### ISSUE TITLE
Error: attempt to call field 'get_node_range' (a nil value)

### ISSUE THREAD

**Issue by mmirus** [AckslD/nvim-FeMaco.lua#13]
Hey there! This is a really cool plugin. Thank you for sharing it!

When I try to use it by running `:FeMaco` with the cursor in a code block, I get this error:

```
Error executing Lua callback: ...te/pack/packer/start/nvim-FeMaco.lua/lua/femaco/edit.lua:33: attempt to call field 'get_node
_range' (a nil value)                                                                                                        
stack traceback:                                                                                                             
        ...te/pack/packer/start/nvim-FeMaco.lua/lua/femaco/edit.lua:33: in function 'get_match_range'                        
        ...te/pack/packer/start/nvim-FeMaco.lua/lua/femaco/edit.lua:79: in function 'get_match_at_cursor'                    
        ...te/pack/packer/start/nvim-FeMaco.lua/lua/femaco/edit.lua:139: in function 'edit_code_block'                       
        ...te/pack/packer/start/nvim-FeMaco.lua/lua/femaco/init.lua:10: in function <...te/pack/packer/start/nvim-FeMaco.lua/
lua/femaco/init.lua:10>
```

I'm on neovim v0.7.2 on  Linux. I used Packer to install the plugin as shown in the README, and I do have treesitter set up.

Any idea what's going wrong? Happy to provide additional info and help troubleshoot however I can.

https://user-images.githubusercontent.com/524994/191059953-aecd151e-9985-4732-9f34-2ded6ddec352.mp4

**Comment by AckslD** [comment]
Ah, forgot that this would only work for 0.8, I pushed a fix for 0.7 in #14, could you try if it works for you?

**Comment by mmirus** [comment]
Thanks for the quick response! That does fix the error and the floating window pops up as expected.

The language servers don't seem to be able to find the project root, though, so they don't work correctly. Is that expected? Or maybe another 0.7 compat issue?


https://user-images.githubusercontent.com/524994/191064796-fd1828fe-91e3-4324-abd8-37a9c867d672.mp4

**Comment by AckslD** [comment]
This would only work for LSPs that support "single file mode" since the file is a standalone temporary file under `/tmp/`. Or would it be helpful if the temporary file would live somewhere else? Could make that configurable.

**Comment by mmirus** [comment]
Good point. For correct root detection, the temporary file probably would have to be written inside the neovim CWD or alongside the file from which `:FeMaco` was invoked, and that would most likely be a very intrusive and unwanted behavior.

One other option I can think of would be overriding the lspconfig `root_dir` somehow, but that seems likely to bring its own set of challenges and is probably out of scope for a plugin like this--though that's obviously up to you!

To me, this plugin is still a nice enhancement and the limitation on the LSP side doesn't stop it from being useful.

**Comment by AckslD** [comment]
> and that would most likely be a very intrusive and unwanted behavior.

It seems so to me also, but we could expose a `create_temporary_file_name` callback if that's useful.

> One other option I can think of would be overriding the lspconfig root_dir somehow, but that seems likely to bring its own set of challenges and is probably out of scope for a plugin like this

Seems indeed out of scope at first sight, but happy to discuss solutions. Could for example the overriding of `root_dir` be done on the user-side in the `post_open_float` callback?

**Comment by bew** [comment]
> For correct root detection, the temporary file probably would have to be written inside the neovim CWD or alongside the file from which :FeMaco was invoked, and that would most likely be a very intrusive and unwanted behavior.

I think it would still be nice to have the possibility to have that behavior (expose the file creation code as an overridable function?), I'm considering having them named in a way that is matched by my global (not per project) git ignore file (see git config `core.excludesfile`)

**Comment by mmirus** [comment]
> It seems so to me also, but we could expose a `create_temporary_file_name` callback if that's useful.

I'm not sure I would use it, but it seems like you have at least one user who would in @bew. It's really up to you what kind of options you want to have to maintain!

> Seems indeed out of scope at first sight, but happy to discuss solutions. Could for example the overriding of `root_dir` be done on the user-side in the `post_open_float` callback?

This might be possible. I wish I know more about how to interact with the buildin LSP / lspconfig beyond basic config. For example, perhaps users might be able to call `vim.lsp.start` from `post_open_float`.

---

Either way, thank you for the quick fix in #14! :partying_face:

**Comment by AckslD** [comment]
@bew I exposed `create_tmp_filepath` in #15. It also takes the filetype as input if that's useful but can also be ignored to create the path. Let me know if that works as you intended.
